source/common/thread_local/thread_local_impl.cc

```diff
@@ -61,6 +110,40 @@ void InstanceImpl::registerThread(Event::Dispatcher& dispatcher, bool main_threa
   }
 }
 
+// Deletes a Slot if it's recycleable(no on-the-fly callbacks points to it), otherwise puts it into
+// a deferred delete queue, and schedules a cleanup task to collect the Slot once it's recycleable.
+void InstanceImpl::recycle(std::unique_ptr<SlotHolder>&& holder) {
+  if (holder->isRecycleable()) {
+    holder.reset();
+    return;
+  }
+  deferred_deletes_.emplace_back(std::move(holder));
+  scheduleCleanup();
+}
+
+// Cleans up the deferred deletes queue.
+// Removes all the items with ref count 1, and reschedule another cleanup task if there are left
+// un-recycleable items.
+// This way ensures that, unless server shuts down, an enqueued item will be
+// deleted eventually.
+void InstanceImpl::scheduleCleanup() {
+  ASSERT(main_thread_dispatcher_ != nullptr);
+  if (shutdown_) {
+    return;
+  }
+  if (deferred_deletes_.empty()) {
+    return;
+  }
+  main_thread_dispatcher_->post([this]() {
```

Comment:
I feel like this ought to be avoidable, but perhaps I'm missing something.  We're using ref_count_ not as an integer, but just to have an object with a use count right?

If we made refcount_ a class, and had Bookkeeper have a reference to it (along with runOnAllThreads calls), and the refcount destructor did a dispatcher post(delete)) couldn't we avoid this extra layer of clean up complexity? We'd still delete when both the Bookkeeper and all thread references were gone, and we'd still delete in the original thread.

Reply:
So the problem with "posting back to main-thread when the propagation is done", there again is this race-window, where the posted callback in main-thread may gets run first, and then the "copy of lambda" in the worker thread may held the last reference of the refcount_.

IMHO, We should also avoid to move the ownership of a slotImpl or its owner, by posting a callback/complete_cb which shares a portion of the object as well.  That may cause complication when ServerImpl tears down, I went into a crash with that, where I posted shared_ptr<RdsRouteConfigProvider> back to main thread dispatcher queue, during shutdown, the SingletonManager is teared down ahead of the DispatcherImpl in ServerImpl, which includes the RouteConfigProviderManager that's referenced by the RdsRouteCnofigProvider in the dispatcher-msg-queue. 

> The Bookeeper is inspiring WRT extending the ability of Slot.
> 
> I am not sure if the SlotHolder is too complex since it is kind of re-inventing shared_ptr. What's more the Bookeeper combining with SlotHolder is duplicating the deleter of shared_ptr.
> 
> My understanding of the bookkeeper is to track
> 
> * how many runOnAllThread on the fly
> * The last of Bookkeeper destruction and `runOnallThread` should destruct the real slot
> 
> I propose to replace SlotHolder by shared_ptr, and bookeeper should
> 1 ref the shared_ptr until destruction (shared_ptr as a member)
> 2 increase the ref of shared_ptr at runOnAllThread (no more complex than capture it)
> 3. choose one from 3.1 and 3.2
> 3.1 BookKeeper set the deleter of the shared_ptr to moving to defer delete queue
> 3.2 rewrite the BookKeeper so that `runOnAllThread` enforce a oncomplete_cb to decrease the shared_ptr on main thread. See https://github.com/envoyproxy/envoy/pull/7011/files . 3.2 doesn't even need a defer delete queue.
> 
> My proposal not decrease the complexity of SlotHolder.
> With 3.2 Bookkeeper::runOnAllThread could cut the overhead and the complexity of operation on deferred deletion.
> 
> WDYT?



> The Bookeeper is inspiring WRT extending the ability of Slot.
> 
> I am not sure if the SlotHolder is too complex since it is kind of re-inventing shared_ptr. What's more the Bookeeper combining with SlotHolder is duplicating the deleter of shared_ptr.
> 
> My understanding of the bookkeeper is to track
> 
> * how many runOnAllThread on the fly
> * The last of Bookkeeper destruction and `runOnallThread` should destruct the real slot
> 
> I propose to replace SlotHolder by shared_ptr, and bookeeper should
> 1 ref the shared_ptr until destruction (shared_ptr as a member)
> 2 increase the ref of shared_ptr at runOnAllThread (no more complex than capture it)
> 3. choose one from 3.1 and 3.2
> 3.1 BookKeeper set the deleter of the shared_ptr to moving to defer delete queue
> 3.2 rewrite the BookKeeper so that `runOnAllThread` enforce a oncomplete_cb to decrease the shared_ptr on main thread. See https://github.com/envoyproxy/envoy/pull/7011/files . 3.2 doesn't even need a defer delete queue.
> 
> My proposal not decrease the complexity of SlotHolder.
> With 3.2 Bookkeeper::runOnAllThread could cut the overhead and the complexity of operation on deferred deletion.
> 
> WDYT?

Yeah, SlotHolder is basically the obj that holds a slot and its outgoing ref-count. I made a class instead using a shared pointer as I thought it may be more "readable". the ferre

Reply:
Oh, SingletonManager early destruction. With this restriction I don't have a good opinion...

Reply:
Talked w/ Alyssa offline, get the idea now. I believe this matches what lamndai suggested as well.
The "schedule a cleanup callback" logic is removed and each Slot Bookkeeper fires a cleanup callback on the ref-count's destruction.